test/e2e/density.go

```diff
@@ -150,11 +154,24 @@ var _ = Describe("Density", func() {
 			}
 			Logf("Found %d events", current)
 
-			// Verify there were no pod killings or failures
-			By("Verifying there were no pod killings or failures")
+			// Verify there were no failures.
+			hist := make(map[string]map[string]int)
 			for _, e := range events {
-				for _, s := range []string{"kill", "fail"} {
-					Expect(e.Reason).NotTo(ContainSubstring(s), "event:' %s', reason: '%s', message: '%s', field path: '%s'", e, e.ObjectMeta.Name, e.Message, e.InvolvedObject.FieldPath)
+				if _, found := hist[e.Reason]; !found {
+					hist[e.Reason] = make(map[string]int)
+				}
+				hist[e.Reason][e.Source.Component]++
+			}
+			Logf("Events histogram:")
+			for reason, histEvents := range hist {
+				for source, number := range histEvents {
+					Logf("- reason: %s, source: %s -> %d", reason, source, number)
+					for _, s := range []string{"kill", "fail"} {
+						if strings.Contains(reason, s) {
+							// TODO: Tune the value of allowed failures.
+							Expect(number).NotTo(BeNumerically(">=", 10), "To many '%s' events from '%s'", reason, source)
```

Comment:
Yeah, this is a good point. How about fail if > 1% of events match the bad patterns? But perhaps should sum all the "bad" events together first before comparing?